### PR TITLE
Feature/min max row

### PR DESCRIPTION
**Pull request by Radabaugh**
Implements #2372

### CODE REVIEW

File: src/web/assets/cp/dist/js/Craft.js [hunk 1]
```diff
@@ -12652,8 +12677,47 @@ Craft.EditableTable = Garnish.Base.extend(
...
 
+            this.updateAddRowButton();
+            // onDeleteRow callback
+            this.settings.onDeleteRow(row.$tr);
+        },
+        canAddRow: function() {
+            return (this.rowCount < this.maxRows);
+        },
         addRow: function() {
+            if (this.hasMaxRows && this.hasMinRows) {
```
Comment (lukeholder) [inline src/web/assets/cp/dist/js/Craft.js]: @Radabaugh not sure if you need to check hasMinRows when you are only looking to add a row at this point

File: src/web/assets/cp/dist/js/Craft.js [hunk 2]
```diff
@@ -12652,8 +12677,47 @@ Craft.EditableTable = Garnish.Base.extend(
                 this.addListener(Garnish.$win, 'resize', 'initializeIfVisible');
             }
         },
+        updateAddRowButton: function() {
+            if ((this.hasMaxRows && this.rowCount >= this.maxRows) ||
+                (this.hasMinRows && this.rowCount < this.minRows)) {
+                this.$addRowBtn.css('opacity', '0.2');
+            } else {
+                this.$addRowBtn.css('opacity', '1');
+            }
```
Comment (lukeholder) [inline src/web/assets/cp/dist/js/Craft.js]: I know the button will no longer add or remove rows if it is greyed out, but other than greying out the button, we will want to also turn off the cursor hand pointer so users no longer think it is clickable.

File: src/fields/Table.php [hunk 3]
```diff
@@ -41,6 +41,18 @@ public static function displayName(): string
     // Properties
     // =========================================================================
 
+    /**
+     * @var string|null Custom add row button label
+     */
+    public $addRowLabel = 'Add a row';
```
Comment (brandonkelly) [inline src/fields/Table.php]: Messages like this should always be run through `Craft::t()`, so the default value can be translated. So leave this value undefined here, and from `init()` add this:

```php
if ($this->addRowLabel === null) {
    $this->addRowLabel = Craft::t('app', 'Add a row');
}
```

Comment (brandonkelly) [reply src/fields/Table.php]: There should also be a blank line between each of these property definitions.

File: src/fields/Table.php [hunk 4]
```diff
@@ -66,10 +78,6 @@ public function init()
     {
         parent::init();
 
-        if ($this->defaults === '') {
-            $this->defaults = [];
-        }
-
```
Comment (brandonkelly) [inline src/fields/Table.php]: Is there a reason this was removed? Assuming it served a purpose.

File: src/fields/Table.php [hunk 5]
```diff
@@ -82,6 +90,20 @@ public function init()
...
+    {
+        $rules = parent::rules();
+        $rules[] = ['minRows', 'compare', 'compareAttribute' => 'maxRows', 'operator' => '<=', 'type' => 'number'];
+        $rules[] = ['maxRows', 'compare', 'compareAttribute' => 'minRows', 'operator' => '>=', 'type' => 'number'];
+        $rules[] = ['minRows', 'integer', 'min' => 0];
+        $rules[] = ['maxRows', 'integer', 'min' => 0];
+        return $rules;
+    }
+
+
```
Comment (brandonkelly) [inline src/fields/Table.php]: Delete one of these blank lines.

File: src/fields/Table.php [hunk 6]
```diff
@@ -160,6 +182,7 @@ public function getSettingsHtml()
             Json::encode($columnSettings, JSON_UNESCAPED_UNICODE).
             ');');
 
+
```
Comment (brandonkelly) [inline src/fields/Table.php]: Delete extra blank line

File: src/fields/Table.php [hunk 7]
```diff
@@ -197,6 +220,7 @@ public function getSettingsHtml()
      */
     public function getInputHtml($value, ElementInterface $element = null): string
     {
+
```
Comment (brandonkelly) [inline src/fields/Table.php]: Delete extra blank line

File: src/fields/Table.php [hunk 8]
```diff
@@ -305,6 +329,7 @@ public function getStaticHtml($value, ElementInterface $element): string
      *
      * @param string $type The cell type
      * @param mixed $value The cell value
+     *
```
Comment (brandonkelly) [inline src/fields/Table.php]: Delete extra blank line

File: src/fields/Table.php [hunk 9]
```diff
@@ -346,6 +371,7 @@ private function _normalizeCellValue(string $type, $value)
      * @param string $type The cell type
      * @param mixed $value The cell value
      * @param string|null &$error The error text to set on the element
+     *
```
Comment (brandonkelly) [inline src/fields/Table.php]: Delete extra blank line

File: src/fields/Table.php [hunk 10]
```diff
@@ -368,6 +394,7 @@ private function _validateCellValue(string $type, $value, string &$error = null)
      * @param mixed $value
      * @param ElementInterface|null $element
      * @param bool $static
+     *
```
Comment (brandonkelly) [inline src/fields/Table.php]: Delete extra blank line

File: src/templates/_components/fieldtypes/Table/settings.html [hunk 11]
```diff
@@ -1,25 +1,54 @@
-{% import '_includes/forms' as forms %}
 
-<input type="hidden" name="defaults" value="">
+{% import '_includes/forms' as forms %}
 
 {{ columnsField|raw }}
 {{ defaultsField|raw }}
 
+{{ forms.textField({
+label: "Min Rows"|t('app'),
```
Comment (brandonkelly) [inline src/templates/_components/fieldtypes/Table/settings.html]: these object keys should be indented (along with following objects).

File: src/templates/_components/fieldtypes/Table/settings.html [hunk 12]
```diff
@@ -1,25 +1,54 @@
...
+    name: 'columnType',
+    instructions: "The type of column this field should get in the database."|t('app'),
+    options: [
+    { value: 'text', label: 'text (~64KB)' },
+    { value: 'mediumtext', label: 'mediumtext (~16MB)' },
+    ],
+    value: field.columnType,
+    warning: (field.id ? "Changing this may result in data loss."|t('app')),
+    }) }}
+</div>
```
Comment (brandonkelly) [inline src/templates/_components/fieldtypes/Table/settings.html]: Lost an indent level here.

File: src/templates/_includes/forms/editableTable.html [hunk 13]
```diff
@@ -3,19 +3,20 @@
...
-                    {%- if col.info is defined -%}
-                        <span class="info">{{ col.info|md|raw }}</span>
-                    {%- endif -%}
-                </th>
+            <th scope="col" class="{{ col.class ?? '' }}">
+                {%- if col.heading is defined and col.heading %}{{ col.heading }}{% else %}&nbsp;{% endif %}
+                {%- if col.info is defined -%}
+                <span class="info">{{ col.info|md|raw }}</span>
+                {%- endif -%}
+            </th>
```
Comment (brandonkelly) [inline src/templates/_includes/forms/editableTable.html]: Lost an indent level here.

Comment (brandonkelly) [reply src/templates/_includes/forms/editableTable.html]: Lots of other ones down the page too.

File: src/templates/_includes/forms/editableTable.html [hunk 14]
```diff
@@ -39,48 +40,48 @@
...
-                                        name: cellName,
-                                        value: value,
-                                        disabled: static
+                                    name: cellName,
+                                    value: value,
+                                    disabled: static
                                     } only %}
                                 {%- default -%}
-                                    <textarea name="{{ cellName }}" rows="1"{% if static %} disabled{% endif %}{% if col.placeholder is defined %} placeholder="{{ col.placeholder }}"{% endif %}>{{ value }}</textarea>
+                                    <textarea name="{{ cellName }}" rows="1"{% if static %} disabled{% endif %}{% if col.placeholder is defined %} placeholder="{{ col.placeholder }}" {% endif %}>{{ value }}</textarea>
```
Comment (brandonkelly) [inline src/templates/_includes/forms/editableTable.html]: Why the extra space?

File: src/templates/_includes/forms/editableTable.html [hunk 15]
```diff
@@ -101,10 +102,12 @@
     {% set jsId = id|namespaceInputId|e('js') %}
     {% set jsName = name|namespaceInputName|e('js') %}
     {% set jsCols = cols|json_encode %}
+    {% set jsMaxRows = maxRows %}
+    {% set jsMinRows = minRows %}
```
Comment (brandonkelly) [inline src/templates/_includes/forms/editableTable.html]: only set these `js*` values if you have a reason to – notice the other ones either end in `|e('js')` or `|json_encode`, maxing their values safe for JavaScript.

File: src/web/assets/cp/dist/js/Craft.js [hunk 16]
```diff
@@ -12596,17 +12596,35 @@ Craft.EditableTable = Garnish.Base.extend(
...
+        minRows: null,
+        rowCount: 0,
+        hasMaxRows: false,
+        hasMinRows: false,
+
         radioCheckboxes: null,
 
-        init: function(id, baseName, columns, settings) {
+        init: function(id, baseName, columns, settings, maxRows, minRows) {
+
```
Comment (brandonkelly) [inline src/web/assets/cp/dist/js/Craft.js]: extra newline

File: src/web/assets/cp/dist/js/Craft.js [hunk 17]
```diff
@@ -12596,17 +12596,35 @@ Craft.EditableTable = Garnish.Base.extend(
...
+            this.maxRows = maxRows;
+            this.minRows = minRows;
+
+            if (this.maxRows !== null) {
+                this.hasMaxRows = true;
+            }
+
+            if (this.minRows !== null) {
+                this.hasMinRows = true;
+            }
```
Comment (brandonkelly) [inline src/web/assets/cp/dist/js/Craft.js]: For clarity, set `hasMinRows` and `hasMaxRows` set to null up at the top of the file, and try this instead here:

```js
this.hasMinRows = this.minRows !== null;
this.hasMaxRows = this.maxRows !== null;
```

File: src/templates/_includes/forms/editableTable.html (hunk 14 above)
Comment (Radabaugh) [reply src/templates/_includes/forms/editableTable.html]: Php storm said "There should be a space between attribute and previous attribute".

File: src/fields/Table.php [hunk 18]
```diff
@@ -410,7 +441,12 @@ private function _getInputHtml($value, ElementInterface $element = null, bool $s
             'name' => $this->handle,
             'cols' => $this->columns,
             'rows' => $value,
-            'static' => $static
+            'static' => $static,
+            'addRowLabel' => $this->addRowLabel,
```
Comment (brandonkelly) [inline src/fields/Table.php]: this should be run through `Craft::t()`

```php
'addRowLabel' => Craft::t('site', $this->addRowLabel),
```

**Pull request merged**